models/inventory/helpers.ts

```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
     );
   }
 }
+
+export async function validateSerialNo(
+  doc: StockMovement | StockTransfer | Invoice
+) {
+  for (const row of doc.items ?? []) {
+    await validateItemRowSerialNo(row, doc.movementType as string);
+  }
+}
+
+async function validateItemRowSerialNo(
+  doc: StockMovementItem | StockTransferItem | InvoiceItem,
+  movementType: string
+) {
+  const idx = doc.idx ?? 0 + 1;
+  const item = doc.item;
+
+  if (!item) {
+    return;
+  }
+
+  if (doc.parentdoc?.cancelled) {
+    return;
+  }
+
+  const hasSerialNo = await doc.fyo.getValue(
+    ModelNameEnum.Item,
+    item,
+    'hasSerialNo'
+  );
+
+  if (hasSerialNo && !doc.serialNo) {
+    throw new ValidationError(
+      [
+        doc.fyo.t`Serial No not set for row ${idx}.`,
+        doc.fyo.t`Serial No is enabled for Item ${item}`,
+      ].join(' ')
+    );
+  }
+
+  if (!hasSerialNo && doc.serialNo) {
+    throw new ValidationError(
+      [
+        doc.fyo.t`Serial No set for row ${idx}.`,
+        doc.fyo.t`Serial No is not enabled for Item ${item}`,
+      ].join(' ')
+    );
+  }
+
+  if (!hasSerialNo) return;
+
+  const serialNos = getSerialNumbers(doc.serialNo as string);
+
+  if (serialNos.length !== doc.quantity) {
+    throw new ValidationError(
+      t`${doc.quantity!} Serial Numbers required for ${doc.item!}. You have provided ${
+        serialNos.length
+      }.`
+    );
+  }
+
+  for (const serialNo of serialNos) {
+    const { name, status, item } = await doc.fyo.db.get(
+      ModelNameEnum.SerialNo,
+      serialNo,
+      ['name', 'status', 'item']
+    );
+
+    if (movementType == 'MaterialIssue') {
+      await validateSNMaterialIssue(
+        doc,
+        name as string,
+        item as string,
+        serialNo,
+        status as string
+      );
+    }
+
+    if (movementType == 'MaterialReceipt') {
+      await validateSNMaterialReceipt(
+        doc,
+        name as string,
+        serialNo,
+        status as string
+      );
+    }
+
+    if (movementType === 'Shipment') {
+      await validateSNShipment(doc, serialNo);
+    }
+
+    if (doc.parentSchemaName === 'PurchaseReceipt') {
+      await validateSNPurchaseReceipt(
+        doc,
+        name as string,
+        serialNo,
+        status as string
+      );
+    }
+  }
+}
+
+export const getSerialNumbers = (serialNo: string): string[] => {
+  return serialNo ? serialNo.split('\n') : [];
+};
+
+export const updateSerialNoStatus = async (
+  doc: Doc,
+  items: StockMovementItem[] | ShipmentItem[],
+  newStatus: string
+) => {
+  for (const item of items) {
+    const serialNos = getSerialNumbers(item.serialNo!);
```

Comment:
Don't use assertions `!`, I know it's there in many places I'm removing them as and when I come across them.

If it `item.serialNo` is not defined you should be handling it instead of asserting that it is defined. This will lead to errors.